javalib/src/main/scala/java/math/Multiplication.scala

```diff
@@ -0,0 +1,454 @@
+   * <td>b<sub>2</sub>*a<sub>1</sub></td>
+   * <td>b<sub>2</sub>*a<sub>0</sub></td>
+   * </tr>
+   *
+   * <tr>
+   * <td></td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * </tr>
+   *
+   * <tr>
+   *
+   * <td align="center">A*B=R=</td>
+   * <td align="center">r<sub>5</sub></td>
+   * <td align="center">r<sub>4</sub></td>
+   * <td align="center">r<sub>3</sub></td>
+   * <td align="center">r<sub>2</sub></td>
+   * <td align="center">r<sub>1</sub></td>
+   * <td align="center">r<sub>0</sub></td>
+   * <td></td>
+   * </tr>
+   *
+   * </tbody>
+   * </table>
+   *
+   *</tt>
+   *
+   * @param op1 first factor of the multiplication {@code op1 >= 0}
+   * @param op2 second factor of the multiplication {@code op2 >= 0}
+   * @return a {@code BigInteger} of value {@code op1 * op2}
+   */
+  def multiplyPAP(a: BigInteger, b: BigInteger): BigInteger = {
+    val aLen = a.numberLength
+    val bLen = b.numberLength
+    val resLength = aLen + bLen
+    val resSign =
+      if (a.sign != b.sign) -1
+      else 1
+
+    if (resLength == 2) {
+      val v = unsignedMultAddAdd(a.digits(0), b.digits(0), 0, 0)
+      val valueLo = v.toInt
+      val valueHi = (v >>> 32).toInt
+      if (valueHi == 0) new BigInteger(resSign, valueLo)
+      else new BigInteger(resSign, 2, Array(valueLo, valueHi))
+    } else {
+      val aDigits = a.digits
+      val bDigits = b.digits
+      val resDigits = new Array[Int](resLength)
+      multArraysPAP(aDigits, aLen, bDigits, bLen, resDigits)
+      val result = new BigInteger(resSign, resLength, resDigits)
+      result.cutOffLeadingZeroes()
+      result
+    }
+  }
+
+  def pow(base: BigInteger, exponent: Int): BigInteger = {
+    var _exponent = exponent
+    var res = BigInteger.ONE
+    var acc = base
+    while (_exponent > 1) {
+      if ((_exponent & 1) != 0)
+        res = res.multiply(acc)
+      acc = {
+        if (acc.numberLength == 1) {
+          acc.multiply(acc)
+        } else {
+          val a = new Array[Int](acc.numberLength << 1)
+          val sq = square(acc.digits, acc.numberLength, a)
+          new BigInteger(1, sq)
+        }
+      }
+      _exponent >>= 1
+    }
+    res = res.multiply(acc)
+    res
+  }
+  
+  /**
+   * It calculates a power of ten, which exponent could be out of 32-bit range.
+   * Note that internally this method will be used in the worst case with
+   * an exponent equals to: {@code Integer.MAX_VALUE - Integer.MIN_VALUE}.
+   * @param exp the exponent of power of ten, it must be positive.
+   * @return a {@code BigInteger} with value {@code 10<sup>exp</sup>}.
+   */
+  def powerOf10(exp: Long): BigInteger = {
+    // "SMALL POWERS"
+    if (exp < bigTenPows.length) {
+      bigTenPows(exp.toInt)
+    }
+    else if (exp <= 50){
+      BigInteger.TEN.pow(exp.toInt)
+    }
+    // "LARGE POWERS"
+    else if (exp <= java.lang.Integer.MAX_VALUE) {
```

Comment:
I now realize that you use `Integer.MAX_VALUE` and `Integer.MIN_VALUE` a lot. It would be better to use the Scala idiomatic variant, i.e., `Int.MaxValue` and `Int.MinValue`.